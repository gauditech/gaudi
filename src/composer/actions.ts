import _ from "lodash";

import { FilteredByKind } from "@src/common/kindFilter";
import {
  assertUnreachable,
  ensureEmpty,
  ensureEqual,
  ensureExists,
  resolveItems,
} from "@src/common/utils";
import { getTypeModel } from "@src/compiler/ast/type";
import { composeValidators, validateFieldType } from "@src/composer/models";
import { composeQuery } from "@src/composer/query";
import { getTypedLiteralValue, refKeyFromRef } from "@src/composer/utils";
import {
  ActionDef,
  ActionHookDef,
  ChangesetDef,
  ChangesetOperationDef,
  CreateOneAction,
  DeleteOneAction,
  ExecuteHookAction,
  FetchOneAction,
  FieldSetter,
  FunctionName,
  QueryDef,
  UpdateOneAction,
} from "@src/types/definition";
import * as Spec from "@src/types/specification";

/**
 * Composes the custom actions block for an endpoint. Adds a default action
 * based on `endpoint.kind` if one is not defined in blueprint.
 * Requires `targets` to construct an initial variable context.
 */
export function composeActionBlock(specs: Spec.Action[]): ActionDef[] {
  // Collect actions from the spec, updating the context during the pass through.
  const actions = specs.map((atom) => {
    switch (atom.kind) {
      case "create":
      case "update": {
        return composeModelAction(atom);
      }
      case "delete": {
        return composeDeleteAction(atom);
      }
      case "execute": {
        return composeExecuteAction(atom);
      }
      case "fetch": {
        return composeFetchAction(atom);
      }
    }
  });

  return actions;
}

function composeDeleteAction(spec: FilteredByKind<Spec.Action, "delete">): DeleteOneAction {
  return {
    kind: "delete-one",
    targetPath: spec.targetPath.map((i) => i.text),
    model: findChangesetModel(spec.targetPath),
  };
}

function composeFetchAction(spec: FilteredByKind<Spec.Action, "fetch">): FetchOneAction {
  const changeset = spec.atoms.map((atom) => atomToChangesetOperation(atom, [], []));
  // fetch action's model is derived from it's query
  const query = composeQuery(spec.query);
  return {
    kind: "fetch-one",
    alias: spec.alias,
    changeset,
    model: query.retType,
    query: query,
  };
}

function composeExecuteAction(spec: FilteredByKind<Spec.Action, "execute">): ExecuteHookAction {
  const changeset = spec.atoms.map((atom) => atomToChangesetOperation(atom, [], []));

  const actionHook: ActionHookDef = {
    args: spec.hook.args.map((arg) => ({
      name: arg.name,
      setter: setterToFieldSetter(arg, changeset),
    })),
    hook: spec.hook.code,
  };

  return {
    kind: "execute-hook",
    hook: actionHook,
    changeset,
    responds: spec.responds,
  };
}

/**
 * Composes a single `ActionDef` based on current variable context, entrypoint, endpoint and action specs.
 */
function composeModelAction(spec: Spec.ModelAction): CreateOneAction | UpdateOneAction {
  const model = findChangesetModel(spec.targetPath);

  // TODO: this should be managed in resolver since it can be a circular reference which must be a compiler error
  const changeset: ChangesetDef = [];
  const resolveResult = resolveItems(
    // atoms to be resolved
    spec.actionAtoms,
    // item name resolver
    (atom) => {
      switch (atom.kind) {
        case "input":
          return atom.target.text;
        case "reference":
          return atom.target.text;
        case "set":
          return atom.target.text;
        case "virtual-input":
          return atom.name;
      }
    },
    // item resolver
    (atom) => {
      const op = atomToChangesetOperation(atom, spec.isPrimary ? [] : [spec.alias], changeset);
      // Add the changeset operation only if not added before
      if (!_.find(changeset, { name: op.name })) {
        changeset.push(op);
      }
    }
  );
  // handle error
  if (resolveResult.kind === "error") {
    console.log(
      "ERRORS",
      resolveResult.errors.map((e) => `${e.name} [${e.error.message ?? e.error}]`)
    );

    throw new Error(
      `Couldn't resolve all field setters: ${resolveResult.errors.map((i) => i.name).join()}`
    );
  }

  // Build the desired `ActionDef`.
  return modelActionFromParts(spec, model, changeset);
}

function expandSetterExpression(expr: Spec.Expr, changeset: ChangesetDef): FieldSetter {
  switch (expr.kind) {
    case "literal": {
      return getTypedLiteralValue(expr.literal);
    }
    case "identifier": {
      const [head, ...tail] = expr.identifier;
      const access = tail.map((i) => i.text);
      switch (head.ref.kind) {
        case "auth":
        case "model":
        case "queryTarget":
        case "target":
        case "action":
        case "repeat":
        case "struct": {
          return {
            kind: "reference-value",
            target: {
              alias: head.text,
              access,
            },
          };
        }
        case "authToken": {
          return {
            kind: "request-auth-token",
            access: ["user", "token"],
          };
        }
        case "virtualInput":
        case "modelAtom": {
          // if path has more than 1 element, it can't be a sibling call
          ensureEqual(access.length, 0, `Unexpected nested sibling ${head.text}: ${access}`);
          // check if sibling name is defined in the changeset
          const siblingOp = _.find(changeset, { name: head.text });
          if (siblingOp) {
            return { kind: "changeset-reference", referenceName: head.text };
          } else {
            throw new Error(`Circular reference: ${head.text}`);
          }
        }
        default:
          return assertUnreachable(head.ref);
      }
    }
    case "function": {
      return {
        kind: "function",
        name: expr.name as FunctionName, // FIXME proper validation
        args: expr.args.map((a) => expandSetterExpression(a, changeset)),
      };
    }
  }
}

function setterToChangesetOperation(
  atom: Spec.ActionAtomSet,
  changeset: ChangesetDef
): ChangesetOperationDef {
  return { name: atom.target.text, setter: setterToFieldSetter(atom.set, changeset) };
}

function setterToFieldSetter(
  set: Spec.ActionAtomSetHook | Spec.ActionAtomSetExp | Spec.ActionAtomSetQuery,
  changeset: ChangesetDef
): FieldSetter {
  switch (set.kind) {
    case "hook": {
      const args = set.hook.args.map((arg) => {
        const setter = setterToFieldSetter(arg, changeset);
        return { name: arg.name, setter };
      });
      return { kind: "fieldset-hook", hook: set.hook.code, args };
    }
    case "expression": {
      const exp = set.expr;
      return expandSetterExpression(exp, changeset);
    }
    case "query": {
      return { kind: "query", query: queryFromSpec(set.query) };
    }
  }
}

function atomToChangesetOperation(
  atom: Spec.ModelActionAtom,
  fieldsetNamespace: string[],
  changeset: ChangesetDef
): ChangesetOperationDef {
  switch (atom.kind) {
    case "virtual-input": {
      return {
        name: atom.name,
        setter: {
          kind: "fieldset-virtual-input",
          type: validateFieldType(atom.type),
          required: !atom.optional,
          nullable: atom.nullable,
          fieldsetAccess: [...fieldsetNamespace, atom.name],
          validators: composeValidators(validateFieldType(atom.type), atom.validators),
        },
      };
    }
    case "input": {
      const astType =
        atom.target.type.kind === "nullable" ? atom.target.type.type : atom.target.type;
      ensureEqual(astType.kind, "primitive");
      const type =
        astType.primitiveKind === "string"
          ? "text"
          : astType.primitiveKind === "float"
          ? "integer"
          : astType.primitiveKind;
      return {
        name: atom.target.text,
        setter: {
          kind: "fieldset-input",
          type,
          required: !atom.optional,
          fieldsetAccess: [...fieldsetNamespace, atom.target.text],
        },
      };
    }
    case "reference": {
      return {
        name: atom.target.text,
        setter: {
          kind: "fieldset-reference-input",
          throughRefKey: refKeyFromRef(atom.through.ref),
          fieldsetAccess: [...fieldsetNamespace, `${atom.target.text}_${atom.through.text}`],
        },
      };
    }
    case "set": {
      return setterToChangesetOperation(atom, changeset);
    }
  }
}

/**
 * Returns a model the changeset operates on. Taken from the end of the resolved path
 * which must not end with a `leaf`.
 *
 * FIXME this function is not specific to `changeset`, rename. This may be deprecated
 *       by proposed changes in `getTypedPathFromContext`.
 */
function findChangesetModel(specTargetPath: Spec.IdentifierRef[]): string {
  const modelName = getTypeModel(specTargetPath.at(-1)!.type);
  if (!modelName) return getTypeModel(specTargetPath.at(-2)!.type)!;
  return modelName;
}

/**
 * Constructs an `ActionDef` for a model action.
 */
function modelActionFromParts(
  spec: Spec.ModelAction,
  model: string,
  changeset: ChangesetDef
): CreateOneAction | UpdateOneAction {
  switch (spec.kind) {
    case "create": {
      return {
        kind: "create-one",
        alias: spec.alias,
        changeset,
        targetPath: spec.targetPath.map((i) => i.text),
        model,
        select: [],
        isPrimary: spec.isPrimary,
      };
    }
    case "update": {
      // FIXME update-many when targetKind is model
      return {
        kind: "update-one",
        changeset,
        alias: spec.alias,
        targetPath: spec.targetPath.map((i) => i.text),
        model,
        filter: undefined,
        select: [],
        isPrimary: spec.isPrimary,
      };
    }
  }
}

export function queryFromSpec(qspec: Spec.Query): QueryDef {
  ensureEmpty(qspec.aggregate, "Aggregates are not yet supported in action queries");

  const pathPrefix = _.first(qspec.from);
  ensureExists(pathPrefix, `Action query "fromModel" path is empty ${qspec.from}`);

  return composeQuery(qspec);
}
