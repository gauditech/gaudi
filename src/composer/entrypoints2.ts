import { kindFilter } from "@src/common/kindFilter";
import * as AST from "@src/compiler/ast/ast";
import { getTypeModel } from "@src/compiler/ast/type";
import { SelectDef, SelectItem } from "@src/types/definition";

export function processSelect(
  models: AST.Model[],
  model: string,
  select: AST.Select | undefined,
  parentNamePath: string[]
): SelectDef {
  if (select === undefined) {
    const modelAtoms = models.find((m) => m.name.text === model)?.atoms ?? [];
    const idField: SelectItem = {
      kind: "field",
      name: "id",
      alias: "id",
      namePath: [...parentNamePath, "id"],
      refKey: `${model}.id`,
    };
    const fields = kindFilter(modelAtoms, "field").map(
      (f): SelectItem => ({
        kind: "field",
        name: f.name.text,
        alias: f.name.text,
        namePath: [...parentNamePath, f.name.text],
        refKey: `${model}.${f.name.text}`,
      })
    );
    const references = kindFilter(modelAtoms, "reference").map(
      (r): SelectItem => ({
        kind: "field",
        name: `${r.name.text}_id`,
        alias: `${r.name.text}_id`,
        namePath: [...parentNamePath, `${r.name.text}_id`],
        refKey: `${model}.${r.name.text}_id`,
      })
    );
    return [idField, ...fields, ...references];
  }

  return select.map(({ target, select }): SelectItem => {
    if (target.kind === "long") {
      throw Error("Long select form unsupported composer");
    }
    if (target.name.ref.kind !== "modelAtom") {
      throw new Error("Unexpected select ref");
    }

    const namePath = [...parentNamePath, target.name.ref.name];

    switch (target.name.ref.atomKind) {
      case "field":
      case "computed":
        return {
          kind: target.name.ref.atomKind,
          refKey: `${target.name.ref.model}.${target.name.ref.name}`,
          name: target.name.ref.name,
          alias: target.name.identifier.text,
          namePath,
        };
      case "hook":
        return {
          kind: "model-hook",
          refKey: `${target.name.ref.model}.${target.name.ref.name}`,
          name: target.name.ref.name,
          alias: target.name.identifier.text,
          namePath,
        };
      case "reference":
      case "relation":
      case "query": {
        const targetModel = getTypeModel(target.name.type);
        // if type has no model, it's an aggregate
        if (!targetModel) {
          return {
            kind: "aggregate",
            refKey: `${target.name.ref.model}.${target.name.ref.name}`,
            name: target.name.ref.name,
            alias: target.name.identifier.text,
            namePath,
          };
        }
        return {
          kind: target.name.ref.atomKind,
          name: target.name.ref.name,
          alias: target.name.identifier.text,
          namePath,
          select: processSelect(models, targetModel, select, namePath),
        };
      }
    }
  });
}
