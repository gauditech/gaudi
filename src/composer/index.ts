import fs from "fs";
import Arcade from "../examples/arcade/definition";
import { setupDb } from "./dbSetup";
import { readDefinition } from "./parser";
const definition = readDefinition(Arcade.models);
// ok now we have definition, let's build something with it!
fs.writeFileSync(
  "examples/arcade/_data.json",
  JSON.stringify(definition, null, 2)
);

const dbSetup = setupDb(definition);
fs.writeFileSync("examples/arcade/_db_setup.sql", dbSetup);
