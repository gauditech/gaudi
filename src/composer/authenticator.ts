import _ from "lodash";

import { getRef } from "@src/common/refs";
import { assertUnreachable } from "@src/common/utils";
import { compile } from "@src/compiler/compiler";
import { getInternalExecutionRuntimeName } from "@src/composer/executionRuntimes";
import { parse } from "@src/parser/parser";
import {
  AuthenticatorMethodDef,
  AuthenticatorNamedModelDef,
  Definition,
} from "@src/types/definition";
import {
  AuthenticatorMethodSpec,
  AuthenticatorSpec,
  Specification,
} from "@src/types/specification";

/**
 * Compose authenticator block.
 */
export function composeAuthenticator(def: Definition, spec: AuthenticatorSpec | undefined): void {
  if (spec == undefined) {
    return;
  }

  // hardcoded authenticator name - not exposed through blueprint cause we don't support multiple auth blocks yet
  const name = "Auth";
  const authUserModel = composeTargetModel(def, spec.authUserModelName);
  const accessTokenModel = composeTargetModel(def, spec.accessTokenModelName);
  const method = composeMethod(def, spec.method);

  def.authenticator = {
    name,
    authUserModel,
    accessTokenModel,
    method,
  };
}

function composeTargetModel(def: Definition, modelName: string): AuthenticatorNamedModelDef {
  // get authenticator target model (injected in compiler)
  const model = getRef.model(def, modelName);

  return {
    name: modelName,
    refKey: model.refKey,
  };
}

function composeMethod(
  def: Definition,
  methodSpec: AuthenticatorMethodSpec
): AuthenticatorMethodDef {
  const kind = methodSpec.kind;
  if (kind === "basic") {
    return {
      kind,
    };
  } else {
    assertUnreachable(kind);
  }
}

/**
 * Create authenticator specs.
 */
export function compileAuthenticatorSpec(
  authenticatorSpec: AuthenticatorSpec | undefined
): Specification {
  if (authenticatorSpec == null) {
    return { models: [], entrypoints: [], populators: [], runtimes: [], authenticator: undefined };
  }

  const authUserModelName = authenticatorSpec.authUserModelName;
  const accessTokenModelName = authenticatorSpec.accessTokenModelName;

  const internalExecRuntimeName = getInternalExecutionRuntimeName();

  const AuthenticatorModel = `
    model ${authUserModelName} {
      field name { type text, validate { min 1 } }
      field username { type text, unique, validate { min 8 } }
      field passwordHash { type text, validate { min 8 } }
      relation tokens { from ${accessTokenModelName}, through authUser }
    }
    model ${accessTokenModelName} {
      field token { type text, unique }
      field expiryDate { type text }
      reference authUser { to ${authUserModelName} }
    }

    entrypoint Auth {
      target model ${authUserModelName} as authUser

      // login
      custom endpoint {
        path "login"
        method POST
        cardinality many

        action {
          fetch as existingAuthUser {
            virtual input username { type text }

            query {
              from ${authUserModelName} as a
              filter a.username is username
              limit 1
            }
            // TODO: throw error id user is not resolved
            // currently, "existingAuthUser" ends up empty and "authenticateUser" hook throws error
          }

          execute {
            virtual input password { type text }

            hook {
              arg clearPassword password
              arg hashPassword existingAuthUser.passwordHash
    
              runtime ${internalExecRuntimeName}
              source authenticateUser from "hooks/actions.js"
            }
          }

          // create access token
          create ${accessTokenModelName} as accessToken {
            set token cryptoToken(32)
            set expiryDate stringify(now() + 3600000) // 1 hour
            set authUser existingAuthUser
          }

          // return token
          execute {
            responds

            hook {
              arg status 200
              arg body_token accessToken.token

              runtime ${internalExecRuntimeName}
              source sendResponse from "hooks/actions.js"
            }
          }
        }
      }

      // logout
      custom endpoint {
        path "logout"
        method POST
        cardinality many

        action {
          fetch ${accessTokenModelName} as accessToken {
 
            query {
              from ${accessTokenModelName}
              filter token is @requestAuthToken
              limit 1
            }
          }

          delete accessToken {}

          // return status
          execute {
            responds

            hook {
              arg status 204

              runtime ${internalExecRuntimeName}
              source sendResponse from "hooks/actions.js"
            }
          }
        }
      }

      // register
      custom endpoint {
        path "register"
        method POST
        cardinality many

        action {
          create authUser {
            input { name, username }
            virtual input password { type text, validate { min 8 } }
            set passwordHash cryptoHash(password, 10)
          }

          // return created user
          execute {
            responds

            hook {
              arg status 201
              // body - see hook for details
              arg body_id authUser.id
              arg body_name authUser.name
              arg body_username authUser.username

              runtime ${internalExecRuntimeName}
              source sendResponse from "hooks/actions.js"
            }
          }
        }
      }
    }
  `;

  return compile(parse(AuthenticatorModel));
}
