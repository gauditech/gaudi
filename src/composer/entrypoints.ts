import _ from "lodash";

import { composeActionBlock } from "./actions";

import { getRef, getTargetModel } from "@src/common/refs";
import { ensureEqual } from "@src/common/utils";
import { uniqueNamePaths } from "@src/runtime/query/build";
import { SelectAST } from "@src/types/ast";
import {
  ActionDef,
  Definition,
  DeleteOneAction,
  EndpointDef,
  EntrypointDef,
  FieldSetterReferenceValue,
  FieldsetDef,
  FieldsetFieldDef,
  ModelDef,
  SelectDef,
  SelectItem,
  TargetDef,
  TargetWithSelectDef,
} from "@src/types/definition";
import { EntrypointSpec } from "@src/types/specification";

export function composeEntrypoints(def: Definition, input: EntrypointSpec[]): void {
  def.entrypoints = input.map((spec) => processEntrypoint(def, spec, []));
}

export type TargetContext = {
  model: ModelDef;
  target: TargetDef;
};

function processEntrypoint(
  def: Definition,
  spec: EntrypointSpec,
  parents: TargetContext[]
): EntrypointDef {
  const target = calculateTarget(
    def,
    parents,
    spec.target.identifier,
    spec.target.alias ?? null,
    spec.identify || "id"
  );
  const name = spec.name;
  const targetModel = getRef.model(def, target.retType);

  const thisContext: TargetContext = { model: targetModel, target };
  const targetParents = [...parents, thisContext];

  return {
    name,
    target,
    endpoints: processEndpoints(def, targetParents, spec),
    entrypoints: spec.entrypoints.map((ispec) => processEntrypoint(def, ispec, targetParents)),
  };
}

export function calculateTarget(
  def: Definition,
  parents: TargetContext[],
  name: string,
  alias: string | null,
  identify: string
): TargetDef {
  const ctxModel = _.last(parents)?.model ?? null;
  const namePath = [...parents.map((p) => p.target.name), name];
  if (ctxModel) {
    const prop = getRef(def, ctxModel.name, name);
    switch (prop.kind) {
      case "reference": {
        const reference = prop;
        const model = getRef.model(def, reference.toModelRefKey);
        return {
          kind: "reference",
          name,
          namePath,
          retType: reference.toModelRefKey,
          refKey: reference.refKey,
          identifyWith: calculateIdentifyWith(def, model, identify),
          alias: alias || `$target_${parents.length}`,
        };
      }
      case "relation": {
        const relation = prop;
        const model = getRef.model(def, relation.fromModelRefKey);
        return {
          kind: "relation",
          name,
          namePath,
          retType: relation.fromModel,
          refKey: relation.refKey,
          identifyWith: calculateIdentifyWith(def, model, identify),
          alias: alias || `$target_${parents.length}`,
        };
      }
      case "query": {
        const query = prop;
        const model = getRef.model(def, query.retType);
        return {
          kind: "query",
          name,
          namePath,
          retType: query.retType,
          refKey: query.refKey,
          identifyWith: calculateIdentifyWith(def, model, identify),
          alias: alias || `$target_${parents.length}`,
        };
      }
      default: {
        throw `${prop.kind} is not a valid entrypoint target`;
      }
    }
  } else {
    const model = getRef.model(def, name);
    return {
      kind: "model",
      name,
      namePath,
      refKey: model.refKey,
      retType: model.name,
      identifyWith: calculateIdentifyWith(def, model, identify),
      alias: alias || `$target_${parents.length}`,
    };
  }
}

function calculateIdentifyWith(
  def: Definition,
  model: ModelDef,
  identify: string | undefined
): EntrypointDef["target"]["identifyWith"] {
  const name = identify ?? "id";
  const prop = getRef(def, model.name, name);
  switch (prop.kind) {
    case "field": {
      const field = prop;
      if (field.type === "boolean") {
        throw new Error("Invalid type of identifiyWith - boolean");
      }
      return {
        name,
        type: field.type,
        refKey: field.refKey,
        paramName: `${model.name.toLowerCase()}_${name}`,
      };
    }
    default:
      throw new Error(`Identify with target must be a field`);
  }
}

function processEndpoints(
  def: Definition,
  parents: TargetContext[],
  entrySpec: EntrypointSpec
): EndpointDef[] {
  const context = _.last(parents)!;
  const targets = parents.map((p) => p.target);

  return entrySpec.endpoints.map((endSpec): EndpointDef => {
    const rawActions = composeActionBlock(def, endSpec.action ?? [], targets, endSpec.type);
    const selectDeps = collectActionDeps(def, rawActions);
    const actions = wrapActionsWithSelect(def, rawActions, selectDeps);
    const targetsWithSelect = wrapTargetsWithSelect(def, targets, selectDeps);
    const parentContext = _.initial(targetsWithSelect);
    const target = _.last(targetsWithSelect)!;

    switch (endSpec.type) {
      case "get": {
        return {
          kind: "get",
          response: processSelect(def, context.model, entrySpec.response, context.target.namePath),
          // actions,
          parentContext,
          target,
        };
      }
      case "list": {
        return {
          kind: "list",
          response: processSelect(def, context.model, entrySpec.response, context.target.namePath),
          // actions,
          parentContext,
          target: _.omit(target, "identifyWith"),
        };
      }
      case "create": {
        const fieldset = fieldsetFromActions(def, actions);
        return {
          kind: "create",
          fieldset,
          actions,
          parentContext,
          target: _.omit(target, "identifyWith"),
          response: processSelect(def, context.model, entrySpec.response, context.target.namePath),
        };
      }
      case "update": {
        const fieldset = fieldsetFromActions(def, actions);
        return {
          kind: "update",
          fieldset,
          actions,
          parentContext,
          target: _.first(wrapTargetsWithSelect(def, [target], selectDeps))!,
          response: processSelect(def, context.model, entrySpec.response, context.target.namePath),
        };
      }
      case "delete": {
        return {
          kind: "delete",
          actions,
          parentContext,
          target,
          response: undefined,
        };
      }
    }
  });
}

export function processSelect(
  def: Definition,
  model: ModelDef,
  selectAST: SelectAST | undefined,
  namePath: string[]
): SelectDef {
  if (selectAST === undefined) {
    return model.fields.map((f) => ({
      kind: "field",
      name: f.name,
      alias: f.name,
      namePath: [...namePath, f.name],
      refKey: f.refKey,
    }));
  } else {
    if (selectAST.select === undefined) {
      // throw new Error(`Select block is missing`);
      // for simplicity, we will allow missing nested select blocks
      return model.fields.map((f) => ({
        kind: "field",
        name: f.name,
        alias: f.name,
        namePath: [...namePath, f.name],
        refKey: f.refKey,
      }));
    }
    const s = selectAST.select;

    return Object.keys(selectAST.select).map((name: string): SelectItem => {
      // what is this?
      const ref = getRef.except(def, model.name, name, ["model"]);
      if (ref.kind === "field") {
        ensureEqual(s[name].select, undefined);
        return {
          kind: ref.kind,
          name,
          alias: name,
          namePath: [...namePath, name],
          refKey: ref.refKey,
        };
      } else if (ref.kind === "model-hook") {
        return {
          kind: ref.kind,
          // refKey: ref.refKey,
          name,
          alias: name,
          namePath: [...namePath, name],
          args: ref.args,
          code: ref.code,
        };
      } else if (ref.kind === "computed") {
        return {
          kind: ref.kind,
          refKey: ref.refKey,
          name,
          alias: name,
          namePath: [...namePath, name],
        };
      } else if (ref.kind === "aggregate") {
        return {
          kind: ref.kind,
          refKey: ref.refKey,
          name,
          alias: name,
          namePath: [...namePath, name],
        };
      } else {
        const targetModel = getTargetModel(def, ref.refKey);
        return {
          kind: ref.kind,
          name,
          namePath: [...namePath, name],
          alias: name,
          select: processSelect(def, targetModel, s[name], [...namePath, name]),
        };
      }
    });
  }
}

export function fieldsetFromActions(def: Definition, actions: ActionDef[]): FieldsetDef {
  const fieldsetWithPaths = actions
    .filter((a): a is Exclude<ActionDef, DeleteOneAction> => a.kind !== "delete-one")
    .flatMap((action) => {
      return _.chain(action.changeset)
        .map(({ name, setter }): null | [string[], FieldsetFieldDef] => {
          switch (setter.kind) {
            case "fieldset-input": {
              const field = getRef.field(def, `${action.model}.${name}`);
              return [
                setter.fieldsetAccess,
                {
                  kind: "field",
                  required: setter.required,
                  type: setter.type,
                  nullable: field.nullable,
                  validators: field.validators,
                },
              ];
            }
            case "fieldset-reference-input": {
              const field = getRef.field(def, setter.throughRefKey);
              return [
                setter.fieldsetAccess,
                {
                  kind: "field",
                  required: true, // FIXME
                  nullable: field.nullable,
                  type: field.type,
                  validators: field.validators,
                },
              ];
            }
            default:
              return null;
          }
        })
        .compact()
        .value();
    });

  return collectFieldsetPaths(fieldsetWithPaths);
}

/**
 * Converts a list of single `FieldsetFieldDef`s with their desired paths
 * into a `FieldsetDef` that nests `FieldsetRecordDef`s in order to respect
 * desired access path for each `FieldsetFieldDef`.ß
 */
function collectFieldsetPaths(paths: [string[], FieldsetFieldDef][]): FieldsetDef {
  const uniqueFieldsetPaths = _.uniqWith(paths, _.isEqual);

  const record = _.chain(uniqueFieldsetPaths)
    .map((p) => p[0][0])
    .uniq()
    .map((name) => {
      const relatedPaths = uniqueFieldsetPaths
        .filter((p) => p[0][0] === name)
        .map((p) => [_.tail(p[0]), p[1]] as [string[], FieldsetFieldDef]);
      if (relatedPaths.length === 1 && relatedPaths[0][0].length === 0) {
        // only a leaf node, return fieldset field
        return [name, relatedPaths[0][1]];
      } else if (relatedPaths.every((p) => p[0].length > 0)) {
        // OK, record without faulty leaf nodes
        return [name, collectFieldsetPaths(relatedPaths)];
      } else {
        // leaf node + non-empty node, this is not correct
        throw new Error(`Error in paths: ${paths.map((p) => p[0].join(".")).sort()}`);
      }
    })
    .fromPairs()
    .value();
  return { kind: "record", nullable: false, record };
}

type SelectDep = FieldSetterReferenceValue["target"];
/**
 * Iterates over actions in order to collect, for each of the actions and a default target
 * context variables, which fields are required in the following actions, so that they can
 * be fetched from the database beforehand.
 * Eg. if a `Repo` aliased as `myrepo` requires `myorg.id`, we need to instruct `myorg`
 * context variable to fetch the `id` so it can be referenced later by `myrepo`.
 */
export function collectActionDeps(def: Definition, actions: ActionDef[]): SelectDep[] {
  // collect all update paths
  const nonDeleteActions = actions.filter(
    (a): a is Exclude<ActionDef, DeleteOneAction> => a.kind !== "delete-one"
  );
  const targetPaths = _.chain(nonDeleteActions)
    .flatMap((a) => {
      switch (a.kind) {
        case "create-one": {
          // there's already a parent setter resolving this path, so we can skip it
          return null;
        }
        case "update-one": {
          try {
            // make sure we're not updating a model directly
            getRef.model(def, _.first(a.targetPath)!);
            return null;
          } catch (e) {
            // last item is what's being updated, we need to collect the id
            const [alias, ...access] = [...a.targetPath, "id"];
            return { alias, access };
          }
        }
      }
    })
    .compact()
    .value();
  // collect all targets
  const setterTargets = nonDeleteActions.flatMap((a) => {
    return a.changeset.flatMap(({ setter: operation }) => {
      switch (operation.kind) {
        case "reference-value": {
          return [operation.target];
        }
        case "fieldset-hook": {
          return operation.args.flatMap(({ setter: operation }) =>
            operation.kind === "reference-value" ? [operation.target] : []
          );
        }
        default: {
          return [];
        }
      }
    });
  });
  return [...setterTargets, ...targetPaths];
}

/**
 * Converts `TargetDef`s into `TargetWithSelectDef`s using select deps to resolve each target's `SelectDef`.
 */
function wrapTargetsWithSelect(
  def: Definition,
  targets: TargetDef[],
  deps: SelectDep[]
): TargetWithSelectDef[] {
  return targets.map((target) => {
    const paths = uniqueNamePaths(
      deps.filter((dep) => dep.alias === target.alias).map((dep) => dep.access)
    );
    const model = getRef.model(def, target.retType);
    const select = pathsToSelectDef(def, model, paths, target.namePath);
    return { ...target, select };
  });
}

/**
 * Inserts `ActionDef`s `select` property using select deps to resolve each target's `SelectDef`.
 * FIXME this is confusing because we don't have a special type for ActionDef with(out) select.
 * Prior to calling this function, every `ActionDef` has an empty (`[]`) select property.
 */
export function wrapActionsWithSelect(
  def: Definition,
  actions: ActionDef[],
  deps: SelectDep[]
): ActionDef[] {
  return actions
    .filter((a): a is Exclude<ActionDef, DeleteOneAction> => a.kind !== "delete-one")
    .map((a): ActionDef => {
      const paths = uniqueNamePaths(deps.filter((d) => d.alias === a.alias).map((a) => a.access));
      const model = getRef.model(def, a.model);

      const select = pathsToSelectDef(def, model, paths, [a.alias]);
      return { ...a, select };
    });
}

function pathsToSelectDef(
  def: Definition,
  model: ModelDef,
  paths: string[][],
  namespace: string[]
): SelectDef {
  const direct = _.chain(paths)
    .map((p) => p[0])
    .uniq()
    .value();
  return direct.map((name): SelectItem => {
    // what is name?
    const ref = getRef(def, model.name, name, ["query", "reference", "relation", "field"]);
    const relatedPaths = paths
      .filter((p) => p[0] === name)
      .map(_.tail)
      .filter((p) => p.length > 0);
    switch (ref.kind) {
      case "field": {
        // ensure leaf
        if (relatedPaths.length) {
          throw new Error(`Field path is not root!`);
        }
        return {
          kind: "field",
          alias: name,
          name,
          refKey: ref.refKey,
          namePath: [...namespace, name],
        };
      }
      default: {
        const newModel = getTargetModel(def, ref.refKey);
        return {
          kind: ref.kind,
          alias: name,
          name,
          namePath: [...namespace, name],
          select: pathsToSelectDef(def, newModel, relatedPaths, [...namespace, name]),
        };
      }
    }
  });
}
