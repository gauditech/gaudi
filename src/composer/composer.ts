import { composeEntrypoints } from "./entrypoints";
import { composeModels } from "./models";

import { composePopulators } from "@src/composer/populators";
import { Definition } from "@src/types/definition";
import { Specification } from "@src/types/specification";

export function compose(input: Specification): Definition {
  // let's start with empty definition
  // sub-composers are expected to mutate it
  const def: Definition = { models: [], entrypoints: [], resolveOrder: [], populators: [] };
  composeModels(def, input.models);
  composeEntrypoints(def, input.entrypoints);
  composePopulators(def, input.populators);

  return def;
}
