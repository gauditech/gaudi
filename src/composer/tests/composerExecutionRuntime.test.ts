import { compile } from "@src/compiler/compiler";
import { compose } from "@src/composer/composer";
import { parse } from "@src/parser/parser";

describe("execution runtime composer", () => {
  it("compose execution runtime", () => {
    const bp = `
       runtime MyRuntime1 {
        sourcePath "./some/path/to/file1.js"
      }

      runtime MyRuntime2 {
        sourcePath "./some/path/to/file2.js"
      }
    `;

    const def = compose(compile(parse(bp)));
    const runtimes = def.runtimes;

    expect(runtimes).toMatchSnapshot();
  });
});
