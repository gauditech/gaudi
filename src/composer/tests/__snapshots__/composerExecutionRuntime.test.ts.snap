// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execution runtime composer composes multiple execution runtimes 1`] = `
[
  {
    "default": true,
    "name": "MyRuntime1",
    "sourcePath": "./some/path/to/file1.js",
    "type": "node",
  },
  {
    "default": false,
    "name": "MyRuntime2",
    "sourcePath": "./some/path/to/file2.js",
    "type": "node",
  },
  {
    "default": false,
    "name": "$GAUDI_INTERNAL",
    "sourcePath": "/home/hkasabasic/work/workspaces/gaudi/gaudi/src/composer/hooks",
    "type": "node",
  },
]
`;

exports[`execution runtime composer composes single execution runtime 1`] = `
[
  {
    "default": true,
    "name": "MyRuntime1",
    "sourcePath": "./some/path/to/file1.js",
    "type": "node",
  },
  {
    "default": false,
    "name": "$GAUDI_INTERNAL",
    "sourcePath": "/home/hkasabasic/work/workspaces/gaudi/gaudi/src/composer/hooks",
    "type": "node",
  },
]
`;

exports[`execution runtime composer makes a single runtime the default one 1`] = `
[
  {
    "default": true,
    "name": "MyRuntime1",
    "sourcePath": "./some/path/to/file1.js",
    "type": "node",
  },
  {
    "default": false,
    "name": "$GAUDI_INTERNAL",
    "sourcePath": "/home/hkasabasic/work/workspaces/gaudi/gaudi/src/composer/hooks",
    "type": "node",
  },
]
`;
